server/src/train/trainvehiclelist.cpp

```diff
@@ -89,27 +94,137 @@ TrainVehicleList::TrainVehicleList(Train& train_, std::string_view parentPropert
 }
 
+void TrainVehicleList::load(WorldLoader& loader, const nlohmann::json& data)
+{
+  nlohmann::json state = loader.getState(getObjectId());
+  nlohmann::json objects = state.value("items", nlohmann::json::array());
+  if(!objects.empty())
+  {
+      std::vector<std::shared_ptr<TrainVehicleListItem>> values;
+      for(const auto& object : objects.items())
+      {
+          nlohmann::json itemId = object.value().value("item_id", nlohmann::json());
+          if(itemId.is_number_unsigned())
+          {
+            auto item = std::make_shared<TrainVehicleListItem>(*this, itemId.get<uint32_t>());
+            m_propertyChanged.emplace(item.get(), item->propertyChanged.connect(std::bind(&TrainVehicleList::propertyChanged, this, std::placeholders::_1)));
+            values.emplace_back(item);
+          }
+          else
+              break;
+      }
+      items.load(std::move(values));
+  }
+  SubObject::load(loader, data);
+  rowCountChanged();
+}
+
+void TrainVehicleList::propertyChanged(BaseProperty &property)
+{
+  if(property.name() == "lob")
+      train().updateLength();
+  else if(property.name() == "total_weight")
+      train().updateWeight();
+  else if(property.name() == "speed_max")
+      train().updateSpeedMax();
+
+  if(!m_models.empty() && TrainVehicleListTableModel::isListedProperty(property.name()))
+  {
+    ObjectPtr obj = property.object().shared_from_this();
+    const uint32_t rows = static_cast<uint32_t>(items.size());
+    for(uint32_t row = 0; row < rows; row++)
+    {
+      if(items[row] == obj)
+      {
+        for(auto& model : m_models)
+        {
+          model->propertyChanged(property, row);
+          break;
+        }
+      }
+    }
+  }
+}
+
 TableModelPtr TrainVehicleList::getModel()
 {
-  return std::make_shared<RailVehicleListTableModel>(*this);
+  return std::make_shared<TrainVehicleListTableModel>(*this);
 }
 
-bool TrainVehicleList::isListedProperty(std::string_view name)
+void TrainVehicleList::rowCountChanged()
 {
-  return RailVehicleListTableModel::isListedProperty(name);
+  const auto size = items.size();
+  //length.setValueInternal(static_cast<uint32_t>(size));
+  for(auto& model : m_models)
+      model->setRowCount(static_cast<uint32_t>(size));
 }
 
-void TrainVehicleList::propertyChanged(BaseProperty& property)
+void TrainVehicleList::rowsChanged(uint32_t first, uint32_t last)
 {
-  if(property.name() == "lob")
-    train().updateLength();
-  else if(property.name() == "total_weight")
-    train().updateWeight();
-  else if(property.name() == "speed_max")
-    train().updateSpeedMax();
+  for(auto& model : m_models)
+  {
+      model->rowsChanged(first, last);
+  }
+}
+
+void TrainVehicleList::addObject(std::shared_ptr<RailVehicle> vehicle)
+{
+  std::shared_ptr<TrainVehicleListItem> object = getItemFromVehicle(vehicle);
+  if(object)
+    return; //Vehicle is already in this train
+
+  object = std::make_shared<TrainVehicleListItem>(*this, getItemId());
```

Comment:
A `TrainVehicleListItem` should always have a vehicle, why not add it to the constructor to make it more clear that it is required

Reply:
Should vehicle property be readonly then?
Like to change vehicle you delete and readd the item.

Reply:
Yes, I think that is better. As long as we don't add a *replace vehicle* option it is not required and event then we will problably use `setValueInteral`.